scripts/prebundle.ts

```diff
@@ -0,0 +1,156 @@
+
+interface BuildOptions {
+  cwd: string;
+  input: string;
+  externals: string[];
+}
+
+async function buildESM({ cwd, input, externals }: BuildOptions) {
+  const bundle = await rollup({
+    input,
+    plugins: [
+      nodeResolve({
+        browser: true,
+        preferBuiltins: true,
+      }),
+      commonjs(),
+      json(),
+      babel({ babelHelpers: 'runtime', skipPreflightCheck: true }),
+      rollupTypescript({ lib: ['es2015', 'dom'], target: 'es6' }),
+      terser({ output: { comments: false }, module: true }),
+    ],
+    external: externals,
+  });
+
+  const previewOutputOptions: OutputOptions = {
+    dir: resolve(cwd, './dist/esm'),
+    format: 'es',
+  };
+
+  await bundle.generate(previewOutputOptions);
+  await bundle.write(previewOutputOptions);
+  await bundle.close();
+}
+
+async function buildModern({ cwd, input, externals }: BuildOptions) {
+  const bundle = await rollup({
+    input,
+    plugins: [
+      nodeResolve({
+        browser: true,
+        preferBuiltins: true,
+      }),
+      commonjs(),
+      json(),
+      babel({
+        babelHelpers: 'runtime',
+        skipPreflightCheck: true,
+        presets: [
+          [
+            '@babel/preset-env',
+            {
+              shippedProposals: true,
+              useBuiltIns: 'usage',
+              corejs: '3',
+              targets: { chrome: '79' },
+            },
+          ],
+        ],
+      }),
+      rollupTypescript({ lib: ['es2015', 'dom'], target: 'es6' }),
+      terser({ output: { comments: false }, module: true }),
+    ],
+    external: externals,
+  });
+
+  const previewOutputOptions: OutputOptions = {
+    dir: resolve(cwd, './dist/modern'),
+    format: 'es',
+  };
+
+  await bundle.generate(previewOutputOptions);
+  await bundle.write(previewOutputOptions);
+  await bundle.close();
+}
+
+async function buildCJS({ cwd, input, externals }: BuildOptions) {
+  const bundle = await rollup({
+    input,
+    plugins: [
+      nodeResolve({
+        browser: false,
+        preferBuiltins: true,
+      }),
+      commonjs(),
+      json(),
+      babel({ babelHelpers: 'runtime', skipPreflightCheck: true }),
+      rollupTypescript({ lib: ['es2015', 'dom'], target: 'es6' }),
+      terser({ output: { comments: false } }),
+    ],
+    external: externals,
+  });
+
+  const previewOutputOptions: OutputOptions = {
+    dir: resolve(cwd, './dist/cjs'),
+    format: 'commonjs',
+  };
+
+  await bundle.generate(previewOutputOptions);
+  await bundle.write(previewOutputOptions);
+  await bundle.close();
+}
+
+async function run() {
+  const cwd = process.cwd();
+  const { packageJson: pkg } = await readPkgUp({ cwd });
+  const input = path.join(cwd, pkg.bundlerEntrypoint);
+  const externals = Object.keys({ ...pkg.dependencies, ...pkg.peerDependencies });
+
+  const options = {
+    cwd,
+    externals,
+    input,
+  };
+
+  await buildESM(options);
+  await buildCJS(options);
+  await buildModern(options);
```

Comment:
Could this happen in parallel? Would that be a bad idea?

Reply:
perhaps I can use a single rollup run to generate all 3..

```
output: [
    {
      format: 'esm',
      dir: 'output',
    },
    {
      format: 'system',
      dir: 'output/legacy',
    },
  ],
  ```